filebeat/crawler/crawler.go

```diff
@@ -6,62 +6,70 @@ import (
 
 	"github.com/elastic/beats/filebeat/input/file"
 	"github.com/elastic/beats/filebeat/prospector"
+	"github.com/elastic/beats/filebeat/registrar"
 	"github.com/elastic/beats/libbeat/common"
 	"github.com/elastic/beats/libbeat/logp"
 )
 
 type Crawler struct {
-	prospectors       []*prospector.Prospector
+	prospectors       map[uint64]*prospector.Prospector
 	prospectorConfigs []*common.Config
 	out               prospector.Outlet
 	wg                sync.WaitGroup
+	once              bool
 }
 
-func New(out prospector.Outlet, prospectorConfigs []*common.Config) (*Crawler, error) {
+func New(out prospector.Outlet, prospectorConfigs []*common.Config, once bool) (*Crawler, error) {
 
 	if len(prospectorConfigs) == 0 {
 		return nil, fmt.Errorf("No prospectors defined. You must have at least one prospector defined in the config file.")
 	}
 
 	return &Crawler{
 		out:               out,
+		prospectors:       map[uint64]*prospector.Prospector{},
 		prospectorConfigs: prospectorConfigs,
+		once:              once,
 	}, nil
 }
 
-func (c *Crawler) Start(states file.States, once bool) error {
+func (c *Crawler) Start(r *registrar.Registrar) error {
 
 	logp.Info("Loading Prospectors: %v", len(c.prospectorConfigs))
 
 	// Prospect the globs/paths given on the command line and launch harvesters
 	for _, prospectorConfig := range c.prospectorConfigs {
-
-		prospector, err := prospector.NewProspector(prospectorConfig, states, c.out)
+		err := c.startProspector(prospectorConfig, r.GetStates())
```

Comment:
Is your concern, that a prospector is stopped which handled the file `a.log` and 2 seconds later an other prospector is started that handles `a.log`? You are right that this could become an issue, even with a central registry for all harvesters, as the harvester could already be finished but the registry is not updated yet. I think a potential solution here is, that as soon as we have a registry, the harvester add itself to the registry and in the registry harvesters are removed then `Finished` is set to true. So as long as the data for the completion of the harvester hasn't arrived at the registry, no new harvester can be started.

Are we on the same page that for this PR this has not effect yet, right?